doc/dev/rule-streaming.md

```diff
@@ -0,0 +1,241 @@
+between directory. In order to calculate the targets produced by
+`copy_files`, which needs to be known at rule production time, we need
+to evaluate the glob given to `copy_files`. Which requires doing a
+call to `Load_rules.load_dir` as previously described. Contrary to the
+other form of dependency we just describe, this ones can go from any
+directory to any other directory. For instance, the following stanza
+in `src/dune`:
+
+```
+(copy_files foo/*.txt)
+```
+
+would create a dependency from `_build/default/src` to
+ `_build_default/src/foo`.
+
+So in the end, if we were looking at the internal computation graph of
+Dune and narrowing it to just the calls to `Load_rules.load_dir`, we
+would see a graph with many edges going from a directory to one of its
+ancestor. These would mostly be between generated dot-subdirectories
+and their first ancestor that has a corresponding directory in the
+source tree. Plus a few other arbitrary ones for each `copy_files`
+stanza.
+
+## Directory targets
+
+Before directory targets, answering the question "what rules produces
+file X?" was easy. Dune would just call `Load_rules.load_dir` and
+lookup `X` in the result. With directory targets, things are a bit
+more complicated. Indeed, `X` might also be produced by a directory
+target in an ancestor directory. This means that `Load_rules.load_dir`
+now need to look in parent directories as well, which introduce more
+dependencies from directories to their parents and can create cycles
+because of `copy_files` stanza that create dependencies in the other
+direction.
+
+At a result, some combination of `copy_files` and directory targets
+don't produce the expected result. This is documented in the test
+suite.
+
+## Problem
+
+The above description expose a concrete problem with directory
+targets, but there is also a general sense of messiness in the way
+things work. Generating rules for multiple directories at once is
+natural, but the current encoding is odd.
+
+## Proposal
+
+The proposal is to add the following rule: `gen_rules ~dir` is allowed
+to produced rules in `dir` or any of its descendant only. It is not
+allowed to produce rules anywhere else.
+
+`Load_rules.load_dir ~dir` will then always call itself recursively on
+the parent of `dir` and take the union of the rules produced by
+`gen_rules` for `dir` and the ones produced by the recursive
+call. `gen_rules` will no longer have to redirect a call via
+`Load_rules.load_dir_and_produce_its_rules`, which we would simply
+remove.
+
+This introduce a cycle with all `copy_files` stanza that copy files
+from a sub-directory. We propose the break this cycle by introducing
+laziness in the rule production code.
+
+### Generating rules with a mask
+
+The idea is that when we produce rules, we will producing rules under
+a current active "mask" that tells us where we are allowed to generate
+files or directories.  Trying to produce a rule with targets not
+matched by this mask will be a runtime error.
```

Comment:
Don't we already have schemes for a similar use case? Would it be possible to unify the two?

Reply:
Absolutely. Scheme is more the implementation of such a thing, i.e. it is what we would use inside `Load_rules`. It is not the API that `dune_rules` will use. So `Scheme` would move to `dune_engine` and we would stop using it in `dune_rules`.